security-manager/src/main/java/org/wildfly/extension/security/manager/deployment/PermissionsParser.java

```diff
@@ -48,8 +51,39 @@
  */
 public class PermissionsParser {
 
+    /**
+     * This method is equivalent to calling {@link #parse(XMLStreamReader, ModuleLoader, ModuleIdentifier, BiConsumer)} with {@code null}
+     * as the last argument.
+     *
+     * @param reader     The {@link XMLStreamReader} for the XML file
+     * @param loader     A {@link ModuleLoader} which will be used to load the configured permission class and create a {@link PermissionFactory}
+     *                   out of it
+     * @param identifier A {@link ModuleIdentifier} which identifies the module, which contains the permission class to load
+     * @return A list of {@link PermissionFactory}s constructed out of the parsed content
+     * @throws XMLStreamException
+     * @deprecated Use {@link #parse(XMLStreamReader, ModuleLoader, ModuleIdentifier, BiConsumer)} instead
+     */
+    @Deprecated
     public static List<PermissionFactory> parse(final XMLStreamReader reader, final ModuleLoader loader, final ModuleIdentifier identifier)
             throws XMLStreamException {
+        return parse(reader, loader, identifier, null);
+    }
+
+    /**
+     * Parses the contents of a permissions.xml/jboss-permissions.xml file
+     *
+     * @param reader             The {@link XMLStreamReader} for the XML file
+     * @param loader             A {@link ModuleLoader} which will be used to load the configured permission class and create a {@link PermissionFactory}
+     *                           out of it
+     * @param identifier         A {@link ModuleIdentifier} which identifies the module, which contains the permission class to load
+     * @param exprExpandFunction A function which will be used, if provided, to expand any expressions (of the form of {@code ${foobar}})
```

Comment:
Hi @jaikiran, what was the reason for introducing a special parameter for this (particularly if we want to standardize on wildfly-common expression parsing)?

Reply:
Hi @dmlloyd, the sole reason (from what I remember) was to allow the existing implementation in WildFly EE (which is backed by JBoss Metadata) to be plugged into this new scheme of things. Here's what I did in the WildFly side https://github.com/jaikiran/wildfly/commit/c6f5a2f06c886b163db33597664c1089431d4fdf. I didn't want to change the semantics of the property expansion, that's already existing in WildFly EE descriptors. It's been a while since I submitted this PR, so might be mis-remembering other details, but I'm pretty sure that was the main reason.

Reply:
OK thanks.